extensions/authentication-passport/src/__tests__/acceptance/fixtures/oauth2-provider.ts

```diff
@@ -0,0 +1,253 @@
+    "2222": {"client_secret":"app2_secret", tokens:{}}
+};
+
+/**
+ * user registry
+ */
+const users = [
+    {id: 1001, username: "user1", password: "abc", email: "usr1@lb.com", signingKey: "AZeb=="},
+    {id: 1002, username: "user2", password: "xyz", email: "usr2@lb2.com", signingKey: "BuIx=+"}
+];
+
+/**
+ * find a user by a name and password
+ * @param {*} username 
+ * @param {*} password 
+ */
+function findUser(username: string, password: string) {
+    const usr = _.filter(users, (user: MyUser) => (user.username === username && user.password === password));
+    if (usr.length > 0) return usr[0];
+    return null;
+}
+
+/**
+ * create a jwt token
+ * @param {*} user 
+ * @param {*} scopes 
+ * @param {*} signingKey 
+ */
+async function createJwt(user: MyUser, scopes: string, signingKey: string, client_id: string) {
+    const jti = Math.floor(Math.random() * Math.floor(1000));
+    const token = await jwt.sign({
+        jti: jti,
+        sub: user.id,
+        name: user.username,
+        email: user.email,
+        iss: 'sample oauth provider',
+        exp: Math.floor(Date.now() / 1000) + (5 * 1000),
+        iat: Math.floor(Date.now() / 1000),
+        grant_type: 'auth code',
+        scopes: scopes,
+        client_id: client_id
+    }, signingKey);
+    return {token: token, id: jti}
+}
+
+/**
+ * verify token
+ * 
+ * check with given client id and token if token is valid
+ * 
+ * @param {*} req 
+ * @param {*} token 
+ */
+async function verifyToken(token: string) {
+    const unwrappedJwt: JWT = jwt.decode(token, {json: true, complete: true});
+    const tokenId: string = unwrappedJwt.payload.jti;
+    const app: App = registeredApps[unwrappedJwt.payload.client_id];
+    if (app?.[tokenId]) {
+        try {
+            const result = await jwt.verify(token, app[tokenId].signingKey);
+            if (result) {
+                return result;
+            } else {
+                throw new Error('invalid token');
+            }
+        } catch (err) {
+            throw err;
+        }
+    } else {
+        throw new Error('invalid app');
+    }
+}
+
+/**
+ * Authorization call: this is an endpoint to authorize registered apps
+ * 
+ * 1. redirects to login page if the client_id is registered
+ * 2. returns error if client_id is not registered
+ */
+app.get("/oauth/dialog", function (req, res) {
+    if (!req.query.redirect_uri) {
+        res.setHeader('Content-Type', 'application/json');
+        res.status(500).send(JSON.stringify({error:  "redirect_uri not sent in query"}));
+    }
+    if (registeredApps[req.query.client_id]) {
+        let params = '?client_id=' + req.query.client_id + '&&redirect_uri=' + req.query.redirect_uri;
+        params = params + '&&scope=' + req.query.scope;
+        res.redirect('/login' + params);
+    } else {
+        res.send("invalid app");
+    }
+});
+
+/**
+ * login page
+ * 
+ * handles login part of the authorization call
+ */
+app.get('/login', function(req, response) {
```

Comment:
Is this code attempting to simulate an oauth2 provider like google and facebook, and you are just using it as a stub for your mocha test cases? Please add some comments to explain the purpose of this file and its methods. thank you.

Reply:
yes exactly this is a stub for a social media app, in fact it works with your express facebook example, https://github.com/emonddr/express-4.x-facebook-example , when I replaced all urls. You can start both apps and login redirects to the mock oauth2 provider, enter user id , password and then see it redirected to your app with an oauth token.

Reply:
I will add necessary comments to make this clear

Reply:
> yes exactly this is a stub for a social media app, in fact it works with your express facebook example, https://github.com/emonddr/express-4.x-facebook-example , when I replaced all urls. You can start both apps and login redirects to the mock oauth2 provider, enter user id , password and then see it redirected to your app with an oauth token.

haha cool. How did you find my app?

Reply:
you had commented in issue https://github.com/strongloop/loopback-next/issues/2624 about your poc.